runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillTimerInternals.java

```diff
@@ -94,8 +94,15 @@ public void setTimer(TimerData timerKey) {
 
   @Override
   public void setTimer(
-      StateNamespace namespace, String timerId, Instant timestamp, TimeDomain timeDomain) {
-    timers.put(timerId, namespace, TimerData.of(timerId, namespace, timestamp, timeDomain));
+      StateNamespace namespace,
+      String timerId,
+      Instant timestamp,
+      Instant outputTimestamp,
+      TimeDomain timeDomain) {
+    timers.put(
+        timerId,
+        namespace,
+        TimerData.of(timerId, namespace, timestamp, outputTimestamp, timeDomain));
     timerStillPresent.put(timerId, namespace, true);
   }
 
```

Comment:
Watermark holds for Dataflow should be set here. Look at the persistTo method in this class which sets and deletes the required holds.